src/typechecker.erl

```diff
@@ -1127,7 +1129,8 @@ expect_record_type(Record, {type, _, union, Tys}, TEnv) ->
     expect_record_union(Record, Tys, TEnv);
 expect_record_type(Record, {var, _, Var}, _TEnv) ->
     {ok, constraints:add_var(Var,
-           constraints:upper(Var, {record, erl_anno:new(0), Record}))};
+           constraints:upper(Var, ?annotate_type({record, erl_anno:new(0), Record},
+                                                 type())))};
```

Comment:
A record type is written `{type,_,record,[{atom,_,my_record}]}`. That's the error.

(The reason `?annotate_type` passes without error is that the literal is considered untyped, which makes it possible to hide the error like this. With infer, a type error would be reported for `?annotate_type`, I think.)